rasa/core/lock_store.py

```diff
@@ -0,0 +1,262 @@
+
+logger = logging.getLogger(__name__)
+
+ACCEPTED_LOCK_STORES = ["in_memory", "redis"]
+
+
+# noinspection PyUnresolvedReferences
+class LockError(Exception):
+    """Exception that is raised when a lock cannot be acquired.
+
+     Attributes:
+          message (str): explanation of which `conversation_id` raised the error
+    """
+
+    pass
+
+
+class LockStore:
+    def __init__(self, lifetime: int = DEFAULT_LOCK_LIFETIME):
+        self.lifetime = lifetime
+
+    @staticmethod
+    def find_lock_store(store: EndpointConfig = None) -> "LockStore":
+        if store is None or store.type is None or store.type == "in_memory":
+            lock_store = InMemoryLockStore()
+        elif store.type == "redis":
+            lock_store = RedisLockStore(host=store.url, **store.kwargs)
+        else:
+            logger.debug(
+                "Could not load built-in `LockStore`, which needs to be of "
+                "type: {}. Trying to load `LockStore` from module path '{}' "
+                "instead."
+                "".format(store.type, ", ".join(ACCEPTED_LOCK_STORES), store.type)
+            )
+            lock_store = LockStore.load_lock_store_from_module_path(store.type)
+
+        logger.debug(
+            "Connected to lock store '{}'.".format(lock_store.__class__.__name__)
+        )
+
+        return lock_store
+
+    @staticmethod
+    def load_lock_store_from_module_path(module_path: Text) -> "LockStore":
+        """Given the name of a `LockStore` module tries to retrieve it."""
+
+        from rasa.utils.common import class_from_module_path
+
+        try:
+            return class_from_module_path(module_path)
+        except ImportError:
+            raise ImportError(
+                "Cannot retrieve `LockStore` from path '{}'.".format(module_path)
+            )
+
+    def create_lock(self, conversation_id: Text) -> TicketLock:
+        """Create and save a new `TicketLock` for `conversation_id`."""
+
+        lock = TicketLock(conversation_id)
+        self.save_lock(lock)
+        return lock
+
+    def get_lock(self, conversation_id: Text) -> Optional[TicketLock]:
+        """Fetch lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def delete_lock(self, conversation_id: Text) -> None:
+        """Delete lock for `conversation_id` from storage."""
+
+        raise NotImplementedError
+
+    def save_lock(self, lock: TicketLock) -> None:
+        """Commit `lock` to storage."""
+
+        raise NotImplementedError
+
+    def issue_ticket(
+        self, conversation_id: Text, lifetime: Union[float, int] = DEFAULT_LOCK_LIFETIME
+    ) -> int:
+        """Issue new ticket for lock associated with `conversation_id`.
+
+        Creates a new lock if none is found.
+        """
+
+        lock = self.get_or_create_lock(conversation_id)
+
+        lifetime = lifetime if lifetime is not None else self.lifetime
+        ticket = lock.issue_ticket(lifetime)
```

Comment:
this is far from atomic. I think it is fine, as we probably will never encounter a case where we are trying to acquire a lock for the same conversation twice so quickly that this will be a problem. nevertheless, we should think about the case where two instances try to acquire a lock for the same conversation in very close succession. what will happen?

Reply:
yeah fair point. I just ran a quick check and this operation took ~2 ms
`ticket issuing took 2.09e-05 seconds`. we can't really "lock" it faster than that can we?